### PR TITLE
Restore

### PR DESCRIPTION
**Pull request by snopoke**
- add support for restore
- add support for ownership cleanliness

### CODE REVIEW

File: prototype/mobile_endpoint/dao.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 from sqlalchemy.orm import contains_eager, defer
```
Comment (snopoke) [inline prototype/mobile_endpoint/dao.py:7]: review file

File: prototype/mobile_endpoint/dao.py [hunk 2]
```diff
@@ -91,11 +118,26 @@ def get_indices():
...
+
+        db.session.commit()
+
+    def commit_restore(self, restore_state):
+        synclog_generic = restore_state.current_sync_log
+        if synclog_generic:
+            _, synclog = Synclog.from_generic(synclog_generic)
+            db.session.add(synclog)
+
         db.session.commit()
```
Comment (snopoke) [inline prototype/mobile_endpoint/dao.py:132]: This will also commit any `OwnershipCleanlinessFlags`

File: prototype/mobile_endpoint/models.py [hunk 3]
```diff
@@ -300,3 +304,38 @@ def from_generic(cls, generic, **kwargs):
         self.index_tree = generic.index_tree.indices
         self.hash = generic.get_state_hash().hash
         return new, self
+
+
+class OwnershipCleanlinessFlag(db.Model):
```
Comment (snopoke) [inline prototype/mobile_endpoint/models.py:319]: Converted to SQLAlchemy

File: prototype/mobile_endpoint/models.py [hunk 4]
```diff
@@ -300,3 +304,38 @@ def from_generic(cls, generic, **kwargs):
...
+                    setattr(instance, field, value)
+            db.session.add(instance)
+            db.session.flush()
+            return instance
+
+
+@event.listens_for(OwnershipCleanlinessFlag, "before_update")
+def gen_default(mapper, connection, instance):
+    if object_session(instance).is_modified(instance, include_collections=False):
+        instance.last_checked = datetime.utcnow()
```
Comment (snopoke) [inline prototype/mobile_endpoint/models.py:351]: This was being done in the `save` method of the Django model

File: prototype/mobile_endpoint/views/restore.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
+from datetime import date
```
Comment (snopoke) [inline prototype/mobile_endpoint/views/restore.py:1]: review file

**Comment by snopoke** [comment]
@NoahCarnahan @czue I've marked the files for review. Mostly just copied the restore code from HQ and made the DB calls go through the DAO.

I've made it only use the new restore code (clean restore).

Things that don't work:
- its using a dummy user so you have to pass the user_id in via the HTTP request
- stock (obviously)
- fixtures

File: prototype/mobile_endpoint/dao.py [hunk 6]
```diff
@@ -132,6 +174,34 @@ def get_reverse_indexed_cases(self, domain, case_ids):
             .options(
                 contains_eager('indices'),
                 defer(CaseData.case_json)
+        ).all()
+
+    def get_open_case_ids(self, domain, owner_id):
+        return [row[0] for row in CaseData.query.with_entities(CaseData.id).filter(
```
Comment (czue) [inline prototype/mobile_endpoint/dao.py:180]: I'm not familiar enough with sqlalchemy to know whether this syntax is smart enough just just pull IDs from the DB. do you know / is it?

Comment (czue) [inline prototype/mobile_endpoint/dao.py:180]: ah, i garner that's what the `with_entities` does yeah?

Comment (czue) [inline prototype/mobile_endpoint/dao.py:180]: [answered my own question](http://docs.sqlalchemy.org/en/rel_1_0/orm/query.html#sqlalchemy.orm.query.Query.with_entities)

File: prototype/mobile_endpoint/dao.py [hunk 7]
```diff
@@ -132,6 +174,34 @@ def get_reverse_indexed_cases(self, domain, case_ids):
...
+            CaseIndex.case_id.in_(case_ids),
+        )]
+
+    def get_last_modified_dates(self, domain, case_ids):
+        return dict(
+            CaseData.query.with_entities(CaseData.id, CaseData.server_modified_on).filter(
+                CaseData.domain == domain,
+                CaseData.id.in_(case_ids)
+            )
         )
```
Comment (czue) [inline prototype/mobile_endpoint/dao.py:205]: :heart_eyes: seeing these in sql

File: prototype/mobile_endpoint/models.py [hunk 8]
```diff
@@ -300,3 +304,38 @@ def from_generic(cls, generic, **kwargs):
...
+    to that ID.
+
+    We use this field to optimize restores.
+    """
+    __tablename__ = 'ownership_cleanliness_flag'
+    domain = db.Column(db.Text(), primary_key=True)
+    owner_id = db.Column(UUID(), primary_key=True)
+    is_clean = db.Column(db.Boolean(), nullable=False, default=False)
+    last_checked = db.Column(db.DateTime(), nullable=False, default=datetime.utcnow())
+    hint = db.Column(db.Text())
```
Comment (czue) [inline prototype/mobile_endpoint/models.py]: should this be a foreign key to the case table?

File: prototype/mobile_endpoint/dao.py (hunk 6 above)
Comment (snopoke) [inline prototype/mobile_endpoint/dao.py:180]: :+1:

File: prototype/mobile_endpoint/models.py (hunk 8 above)
Comment (snopoke) [inline prototype/mobile_endpoint/models.py]: Yes, good point.

**Comment by czue** [comment]
looks good! i didn't review anything that wasn't tagged very carefully. feel free to self-merge if you want to get it in

**Pull request merged**